runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperPlugin.java

```diff
@@ -0,0 +1,359 @@
+			}
+		}, 0, 30, TimeUnit.MINUTES);
+	}
+
+	@Override
+	protected void shutDown() throws Exception
+	{
+		keyManager.unregisterKeyListener(inputListener);
+
+		lookupFuture.cancel(true);
+	}
+
+	@Subscribe
+	public void onMenuEntryAdded(MenuEntryAdded event)
+	{
+		int groupId = WidgetInfo.TO_GROUP(event.getActionParam1());
+		String option = event.getOption();
+
+		if (groupId == WidgetInfo.FRIENDS_LIST.getGroupId() || groupId == WidgetInfo.CLAN_CHAT.getGroupId() ||
+			groupId == WidgetInfo.CHATBOX.getGroupId() && !KICK_OPTION.equals(option) || //prevent from adding for Kick option (interferes with the raiding party one)
+			groupId == WidgetInfo.RAIDING_PARTY.getGroupId() || groupId == WidgetInfo.PRIVATE_CHAT_MESSAGE.getGroupId())
+		{
+			boolean after;
+
+			if (AFTER_OPTIONS.contains(option))
+			{
+				after = true;
+			}
+			else if (BEFORE_OPTIONS.contains(option))
+			{
+				after = false;
+			}
+			else
+			{
+				return;
+			}
+
+			// Don't add entry if user is offline
+			ChatPlayer player = getChatPlayerFromName(event.getTarget());
+
+			if (player == null || player.getWorld() == 0 || player.getWorld() == client.getWorld())
+			{
+				return;
+			}
+
+			final MenuEntry hopTo = new MenuEntry();
+			hopTo.setOption(HOP_TO);
+			hopTo.setTarget(event.getTarget());
+			hopTo.setType(MenuAction.RUNELITE.getId());
+			hopTo.setParam0(event.getActionParam0());
+			hopTo.setParam1(event.getActionParam1());
+
+			insertMenuEntry(hopTo, client.getMenuEntries(), after);
+		}
+	}
+
+	private void insertMenuEntry(MenuEntry newEntry, MenuEntry[] entries, boolean after)
+	{
+		MenuEntry[] newMenu = ObjectArrays.concat(entries, newEntry);
+
+		if (after)
+		{
+			int menuEntryCount = newMenu.length;
+			ArrayUtils.swap(newMenu, menuEntryCount - 1, menuEntryCount - 2);
+		}
+
+		client.setMenuEntries(newMenu);
+	}
+
+	@Subscribe
+	public void onHopToMenuClicked(PlayerMenuOptionClicked event)
+	{
+		if (!event.getMenuOption().equals(HOP_TO))
+		{
+			return;
+		}
+
+		ChatPlayer player = getChatPlayerFromName(event.getMenuTarget());
+
+		if (player != null)
+		{
+			hop(player.getWorld());
+		}
+	}
+
+	void hop(boolean previous)
+	{
+		if (client.getGameState() != GameState.LOGGED_IN)
+		{
+			return;
+		}
+
+		int totalLevel = client.getTotalLevel();
+
+		World currentWorld = worldResult.findWorld(client.getWorld());
+
+		// Make it so you always hop out of PVP, high risk and bounty worlds
+		// Also don't regard skill total worlds as a type that must be hopped between
+		EnumSet<WorldType> currentWorldTypes = currentWorld.getTypes().clone();
+		currentWorldTypes.remove(WorldType.PVP);
+		currentWorldTypes.remove(WorldType.PVP_HIGH_RISK);
+		currentWorldTypes.remove(WorldType.BOUNTY);
+		currentWorldTypes.remove(WorldType.SKILL_TOTAL);
+
+		int currentWorldIndex = worldResult.getWorlds().indexOf(currentWorld);
+		CyclicListIterator<World> worldIterator = new CyclicListIterator<>(worldResult.getWorlds(), currentWorldIndex);
```

Comment:
This entire iterator is just `current +/-= 1; current = min(worldList.size(), max(0, current));`

Reply:
I think that doesn't work if you're on world 301 and want to go to the previous world. I made a thing that works for both going to the previous from 301 and going to the next from 424. Maybe it can be improved but it works.